test/test_train_mp_mnist.py

```diff
@@ -0,0 +1,160 @@
+                          28), torch.zeros(FLAGS.batch_size,
+                                           dtype=torch.int64)),
+        sample_count=10000 // FLAGS.batch_size // xm.xrt_world_size())
+  else:
+    train_dataset = datasets.MNIST(
+        FLAGS.datadir,
+        train=True,
+        download=True,
+        transform=transforms.Compose(
+            [transforms.ToTensor(),
+             transforms.Normalize((0.1307,), (0.3081,))]))
+    test_dataset = datasets.MNIST(
+        FLAGS.datadir,
+        train=False,
+        transform=transforms.Compose(
+            [transforms.ToTensor(),
+             transforms.Normalize((0.1307,), (0.3081,))]))
+    train_sampler = None
+    if xm.xrt_world_size() > 1:
+      train_sampler = torch.utils.data.distributed.DistributedSampler(
+          train_dataset,
+          num_replicas=xm.xrt_world_size(),
+          rank=xm.get_ordinal(),
+          shuffle=True)
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        batch_size=FLAGS.batch_size,
+        sampler=train_sampler,
+        shuffle=False if train_sampler else True,
+        num_workers=FLAGS.num_workers)
+    test_loader = torch.utils.data.DataLoader(
+        test_dataset,
+        batch_size=FLAGS.batch_size,
+        shuffle=False,
+        num_workers=FLAGS.num_workers)
+
+  # Scale learning rate to num cores
+  lr = FLAGS.lr * xm.xrt_world_size()
+
+  device = xm.xla_device()
+  model = MNIST().to(device)
+  optimizer = optim.SGD(model.parameters(), lr=lr, momentum=FLAGS.momentum)
+  loss_fn = nn.NLLLoss()
+
+  def train_loop_fn(loader):
+    tracker = xm.RateTracker()
+
+    model.train()
+    for x, (data, target) in loader:
+      optimizer.zero_grad()
+      output = model(data)
+      loss = loss_fn(output, target)
+      loss.backward()
+      xm.optimizer_step(optimizer)
+      tracker.add(FLAGS.batch_size)
+      if x % FLAGS.log_steps == 0:
+        test_utils.print_training_update(device, x, loss.item(), tracker.rate(),
+                                         tracker.global_rate())
+
+  def test_loop_fn(loader):
+    total_samples = 0
+    correct = 0
+    model.eval()
+    for x, (data, target) in loader:
+      output = model(data)
+      pred = output.max(1, keepdim=True)[1]
+      correct += pred.eq(target.view_as(pred)).sum().item()
+      total_samples += data.size()[0]
+
+    accuracy = 100.0 * correct / total_samples
+    test_utils.print_test_update(device, accuracy)
+    return accuracy
+
+  accuracy = 0.0
+  for epoch in range(1, FLAGS.num_epochs + 1):
+    para_loader = dp.ParallelLoader(train_loader, [device])
+    train_loop_fn(para_loader.per_device_loader(device))
+
+    para_loader = dp.ParallelLoader(test_loader, [device])
+    accuracy = test_loop_fn(para_loader.per_device_loader(device))
+    if FLAGS.metrics_debug:
+      print(torch_xla._XLAC._xla_metrics_report())
+
+  return accuracy
+
+
+def _mp_fn(index, flags):
+  global FLAGS
+  FLAGS = flags
```

Comment:
what is the point of these 2 lines?

Reply:
It is in the PR comment.
Essentially pytorch forces its multiprocessing "vector" to be "spawn".
This means that there is no `fork()`, but instead a new process is created, and the Python state pickled and sent to the new process.
And `multiprocessing.spawn()` only pickles whatever is passed to the target function.
This means that any changes to the global state before the `spawn()` call, will not be reflected into the child processes.

Reply:
I see. This looks weird, can we then pass FLAGS around as an arg?

```python
def train_mnist(args):
  ... (s/FLAGS/args/g)
```

wdyt?

Reply:
The convention for globals is all uppercase.
I am not sure what you mean with passing around?
Check the documentation of Python multiprocessing WRT the spawn way of creating processes.
This is what pytorch forces into, since I think they are not liking fork().

Reply:
if I understand correctly, the issue here is, `train_mnist` must have access to the FLAGS object, but spawned subprocesses drop it. However, you're allowed to carry things to the subprocess thru the form of function args, so you do

```
global FLAGS
FLAGS = flags
```

to re-make the global FLAGS object in every subprocess, so `train_mnist` doesn't error.

I'm suggesting to add an argument to train_mnist, and drop the dependency to the global variable. That's pretty much what fairseq does and I think it would work here too, no?

Reply:
What is happening is that the main process starts, and goes through the global scope code, which in our cases populates the global FLAGS variable.
That flags variable will not be populated in the child processes since no arguments will be passed. The new process will get a special argument for the pickled data.
Every piece of state which is modified in the master/parent process, which you want carried over to the child, must be a target function argument.

As far as replacing FLAGS with args, maybe not for this PR.
The is some a bit tricky stuff in test_utils.py that needs to be figured out.